src/sentry/web/frontend/oauth_revoke.py

```diff
@@ -0,0 +1,155 @@
+import logging
+from hashlib import sha256
+from typing import Optional
+
+from django.db.models import Q
+from django.http import HttpRequest, HttpResponse
+from django.utils.decorators import method_decorator
+from django.views.decorators.cache import never_cache
+from django.views.decorators.csrf import csrf_exempt
+from django.views.generic.base import View
+
+from sentry.models import ApiApplication, ApiApplicationStatus, ApiToken
+from sentry.utils import json
+
+logger = logging.getLogger("sentry.api.oauth_revoke")
+
+
+class OAuthRevokeView(View):
+    """
+    OAuth 2.0 token revoke endpoint per RFC 7009
+    https://www.rfc-editor.org/rfc/rfc7009
+
+    Clients can provide either the access_token or refresh_token in the request.
+
+    When revoking the token, the associated access_token or refresh_token is also
+    revoked.
+    """
+
+    @csrf_exempt
+    @method_decorator(never_cache)
+    def dispatch(self, request, *args, **kwargs):
+        return super().dispatch(request, *args, **kwargs)
+
+    def error(self, request: HttpRequest, name, error_description=None, status=400):
```

Comment:
Can we add typing to this (and the rest of the methods in here)?

Reply:
Hmm. Added some typing here, but I'm running into this error in the CI. Have any advice on the proper way to fix this?

> src/sentry/web/frontend/oauth_revoke.py:131: error: Item "None" of "Optional[str]" has no attribute "encode"  [union-attr]

```python
        # only delete the token if one was found
        if token_to_delete:
            token_to_delete.delete()
            logger.info(
                "oauth.revoke-success",
                extra={
                    "client_id": client_id,
                    "application_id": application.id,
                    # don't log the actual token, just a hash of it
                    "sha256_provided_token": sha256(token.encode("utf-8")).hexdigest(),
                    "sha256_access_token": sha256(
                        token_to_delete.token.encode("utf-8")
                    ).hexdigest(),
                    "sha256_refresh_token": sha256(
                        token_to_delete.refresh_token.encode("utf-8")
                    ).hexdigest(),
                    "resource_owner_id": token_to_delete.user.id,
                },
            )
```

I have the `if token_to_delete:` in place, but this doesn't seem to be enough to satisfy mypy. I've also tried this with `if isinstance(token_to_delete, ApiToken):` with the same result.

Reply:
Ah, actually I should have paid closer attention. It is not typing error on `ApiToken`, but rather it's `refresh_token` property that can be `None`.